main.tf

```diff
@@ -3,14 +3,29 @@
 #
 # Creates ICD Postgresql instance
 ##############################################################################
+data "ibm_cbr_zone" "cbr_zone" {
+  count   = length(local.network_zone_id)
+  zone_id = local.network_zone_id[count.index]
+}
 
 locals {
   kp_backup_crn        = var.backup_encryption_key_crn != null ? var.backup_encryption_key_crn : var.key_protect_key_crn
   auto_scaling_enabled = var.auto_scaling == null ? [] : [1]
+  network_zone_id      = flatten([for rule in var.cbr_rules : [for contexts in rule.rule_contexts : [for attributes in contexts : [for attribute in attributes : attribute.value if contains(["networkZoneId"], attribute.name)]]]])
+  validate_cbr_zone    = anytrue(flatten([for cbr_zone in data.ibm_cbr_zone.cbr_zone : [for address in cbr_zone.addresses : address.type == "serviceRef" ? true : false]]))
+}
+resource "null_resource" "validate" {
+  depends_on = [data.ibm_cbr_zone.cbr_zone]
+  provisioner "local-exec" {
+    command     = local.validate_cbr_zone ? "echo Cloud Databases does not support service reference in network zone. ; exit 1" : "echo "
+    interpreter = ["/bin/bash", "-c"]
+  }
```

Comment:
> Not sure we should use this for validation. It uses a `null_resource` block and a `depends_on` which are both not recommended by terraform best practises. Is this use case solvable using this approach for example https://github.com/terraform-ibm-modules/terraform-ibm-cos/blob/main/profiles/fscloud/main.tf#L3

@ocofaigh I initially tried using `tobool` and `regex` but variables within local were evaluated earlier than the execution of `data block` which made the result inconsistent as well was not able to catch the expected error. But for catching the error `data block` need to execute earlier then local, So for maintaining the order I have used `null_resource` as well as `depends_on`.

Reply:
If I try to use `tobool `or`regex` for validation,I get this error which is getting catch by provider  instead of getting catch by our added validation .
<img width="1308" alt="Screenshot 2023-04-12 at 12 21 26 PM" src="https://user-images.githubusercontent.com/56439894/231375043-dbc2b58b-377a-45db-b336-e07bd70f2ab3.png">

Reply:
So I guess this is a special case because of the data lookup? I think maybe we can just go with this solution for now as a quick and dirty hack

Reply:
@Khuzaima05 If I remember rightly, I tried tobool in your code and it was working.  R u sure tobool() is also throwing the above error? I can try once again to confirm if you need.

Reply:
@ocofaigh @jojustin 

There are 2 issues in the discussion as -

- `depends_on` : variables within local block were evaluated earlier than the execution of data block which made the result inconsistent as well as it was not able to catch the expected error. But for catching the error data block need to execute earlier then local, So for maintaining the order we have to use  `depends_on` and `null_resource`.

- `tobool` : when we are using `null_resource` , we cannot use validation as [such](https://github.com/terraform-ibm-modules/terraform-ibm-cos/blob/main/profiles/fscloud/main.tf#L3) directly inside `null_resource` block. We have to use either `trigger` or `provisioner "local-exec"`. So here I have use provisioner.

Reply:
`preconditions ` may be a better tool for this use case, provided it gives enough flexibility on the list of datas. See https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#examples